MCPWM 5.0

usbalbin opened the pull request:
The beginning of a very very rough draft of IDF 5.0 style MCPWM. (mostly reused from #93)

Counter proposal for #93

Note: This depends on [esp-idf/esp-idf-sys#133](https://github.com/esp-rs/esp-idf-sys/pull/133)

ivmarkov commented:
@usbalbin @MabezDev Folks, I'm missing a little bit the context of what is going on here. Can somebody summarize it for me? I'm struggling with the following:
- We have #93 now. However it does not explain what it is improving upon? Is it about the fact that you can somehow support two duty signals with it?
- Then we have this one (#132). It is named "MCPWM 5.0". Also no explanations. Is this also supporting two duty signals, but **only** for ESP IDF 5? If so, what would be our story for ESP IDF 4.4, which would be around for quite some time?

@usbalbin I think it would be very useful if you create a comment that summarizes what these changes are about, and what is their relation to ESP IDF4 vs ESP IDF 5. I think you might be assuming that we have a bit more context than what we actually know about this stuff. :) Thanks!

usbalbin commented on examples/mcpwm-simple.rs:
```diff
@@ -70,22 +70,24 @@ fn main() -> anyhow::Result<()> {
+            peripherals.pins.gpio5,
+        )?;
+
+    let (timer, operator, _, _) = timer.split();
```
Here is one of the larger changes from "MCPWM from IDF <= 4.4" to "MCPWM from IDF 5.0". In 4.4, the concept of timer and operator is used interchangeably since every timer is hardcoded to exactly one operator(In #93 I expose the type `Operator` which control both the operator and its timer).

However, in 5.0 the mapping of timer to operator is no longer hardcoded. An operator can have any timer within the same MCPWM unit, thus multiple operators may share the same timer. To describe this, I have in this PR split this into two types `Operator` and `Timer`.

Since an operator needs a timer to work, we need to somehow ensure that the timer will live atleast as long as the operator. This could be done by having `Operator` hold a reference to its `Timer`. That would however prevent useful things like changing the frequency of the timer while an operator holds a reference to it.

(Please let me know if there is a better way/if this would not work)
To get around this problem I created `TimerConnection` which takes ownership of a timer and 0-3 operators. There is no way to move the `Operators` or the `Timer` out of the `TimerConnection` thus they are garantueed to live as the `Timer`. (mem::swap/replace would be one way to do it. However due to all members having unique types, it is impossible to find an instance of the same type to swap with.)

To get access to the objects within `TimerConnection` one has to call `TimerConnection::split()` which returns a tuple with mut references to every member.

usbalbin commented on examples/mcpwm-deadtime.rs:
```diff
@@ -0,0 +1,106 @@
+// TODO: Update those graphs with measured results
```
Oops, this file was not supposed to be added quite yet...

usbalbin commented on .cargo/config.toml:
```diff
@@ -1,22 +1,31 @@
 [target.xtensa-esp32-espidf]
```
TODO: Figure out which (if any) of the canges in this file was actually needed

usbalbin commented on src/mcpwm/comparator.rs:
```diff
@@ -0,0 +1,116 @@
+impl Comparator {
+    pub(crate) unsafe fn configure(&mut self, gen: &mut mcpwm_gen_t, cfg: CountingDirection) {
+        // TODO: "must be terminated by MCPWM_GEN_TIMER_EVENT_ACTION_END()"
+        mcpwm_generator_set_actions_on_compare_event(
```
[`mcpwm_generator_set_actions_on_compare_event`](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/mcpwm.html#_CPPv444mcpwm_generator_set_actions_on_compare_event18mcpwm_gen_handle_t32mcpwm_gen_compare_event_action_tz) is a variadic function that accepts a list of `mcpwm_gen_handle_t generator, mcpwm_gen_compare_event_action_t ev_act` pairs. Terminated by a special set of values.

Since we can only pass one single pair of values from rust this seems impossible, right?(unless the termination values is all we pass, which would be pointless)

usbalbin replied:
Looking at the implementation of [mcpwm_generator_set_actions_on_timer_event](https://github.com/espressif/esp-idf/blob/5f25bd843721dbddb76ab69099dc1b3f4ae5f759/components/driver/mcpwm/mcpwm_gen.c#L163) the functions seems to continue through the arguments until an `ev_act` with an `event = MCPWM_TIMER_EVENT_INVALID` is found. Looks like this is what [MCPWM_GEN_TIMER_EVENT_ACTION_END](https://github.com/espressif/esp-idf/blob/fde4afc67a2035a1f461dc511cd85dde1ef90368/components/driver/include/driver/mcpwm_gen.h#L87)

Same thing for [mcpwm_generator_set_actions_on_compare_event](https://github.com/espressif/esp-idf/blob/5f25bd843721dbddb76ab69099dc1b3f4ae5f759/components/driver/mcpwm/mcpwm_gen.c#L182) and `MCPWM_GEN_COMPARE_EVENT_ACTION_END()` but here a null pointer is used as `comparator`.

Still, we would need to be able to pass atleast two pairs of values(the real value pair and the end marker) to do something useful. I assume this is hard from rust?

usbalbin replied:
I have little to no knowledge on C variadic functions or ABIs, but would calling a function like
```rust
// Also, make sure bindgen does not generate bindings for the single `ec_act` version of this function
extern "C" fn mcpwm_generator_set_actions_on_compare_event(
    gen: mcpwm_gen_handle_t,
    ev_act0: mcpwm_gen_compare_event_action_t,
    ev_act1: mcpwm_gen_compare_event_action_t
) -> esp_err_t
```

just do the right thing and place the values in the places expected by the called function that actually is variadic?

MabezDev replied:
Yes I think it works like that. On the rust side you just call it with however many args you want and the compiler takes care of making sure the variadic ABI is correct.

usbalbin commented:
This is still quite the mess, and there are a lot of things I am unsure of. However before I get too far down the road, may I request a very light review of your thoughts of the overall picture of this PR?

* Does the API make sense?
* Does the usage in the example mcpwm-simple make sense?
* Any ideas as for how to reduce the perhaps a little bit too heavy use of generics :)
* Naming of things...

usbalbin commented:
I just noticed there is https://github.com/esp-rs/esp-hal/pull/255 , perhaps there could be some inspiration to take from there?

MabezDev commented on src/mcpwm/timer_connection.rs:
```diff
@@ -0,0 +1,158 @@
+    O1: OptionalOperator<1, G>,
+    O2: OptionalOperator<2, G>,
+{
+    pub fn attatch_operator0<CMPX, CMPY, GENA, GENB>(
```
So for v4.4, these bounds don't currently stop you from using timer0 & operator2, right? Perhaps we can somehow use const generics here such that TimerConnection<N> must always match Operator<N> for v4.4 :thinking:.

usbalbin replied:
Something roughly like(semi made up code with bad names)?

```rust
struct CanAttach<const TIMER: u8, const OPERATOR: u8>;
trait Yes;

#[cfg(v4_4)]
impl Yes for CanAttach<0, 0>{}

#[cfg(v4_4)]
impl Yes for CanAttach<1, 1>{}

#[cfg(v4_4)]
impl Yes for CanAttach<2, 2>{}

#[cfg(not(v4_4)]
impl<const TIMER: u8, const OPERATOR: u8> Yes for CanAttach<TIMER, OPERATOR>{}
```

```diff
 impl<const N: u8, G, O1, O2> TimerConnection<N, G, NoOperator, O1, O2>
 where
     G: Group,
     O1: OptionalOperator<1, G>,
     O2: OptionalOperator<2, G>,
+    CanAttach<N, 0>: Yes,
 {
     pub fn attatch_operator0<CMPX, CMPY, GENA, GENB>(
```

MabezDev replied:
Yep! This is basically what I had in mind - maybe one day when const expressions are stable we can simply further.

MabezDev reviewed (commented):
The API looks sensible and easy to use. We need to figure out how to support v4.4 as well (a limitation https://github.com/esp-rs/esp-hal/pull/255 isn't bound by), I've dropped a quick comment about that but I'm sure you'll figure out a clean way of managing it :).

Overall I like the direction of this PR, so I think you're safe to move forward.

usbalbin commented on src/mcpwm/operator_config.rs:
```diff
@@ -0,0 +1,151 @@
+impl<CMPX: OptionalCmpCfg, CMPY: OptionalCmpCfg, GENB: OptionalGenCfg>
+    OperatorConfig<CMPX, CMPY, NoGenCfg, GENB>
+{
+    #[allow(clippy::type_complexity)]
+    fn gen_a<P: OutputPin>(
+        self,
+        config: GeneratorConfig<GenA, CMPX::OnMatchCfg, CMPY::OnMatchCfg, P>,
+    ) -> OperatorConfig<
+        CMPX,
+        CMPY,
+        GeneratorConfig<GenA, CMPX::OnMatchCfg, CMPY::OnMatchCfg, P>,
+        GENB,
+    > {
```
Is there any way to make clippy happy about such a type?

MabezDev replied:
I think silencing clippy here is okay :).

usbalbin commented:
Rebased over master and removed duplicate commits from #183

usbalbin commented:
I have now tested `examples/mcpwm-simple.rs` on an ESP32-S3 and it does (after a lot of fixes) seem to work.

usbalbin commented on src/mcpwm/timer.rs:
```diff
@@ -32,8 +34,8 @@ pub struct TimerConfig {
-            resolution: 160.MHz().into(),
-            period_ticks: 16_000, // 10kHz
+            resolution: 80.MHz().into(),
+            period_ticks: 8_000, // 10kHz
```
As far as I can tell the timer group's prescaler is for some reason [hardcoded to `2`](https://github.com/espressif/esp-idf/blob/7052a14d61d98dd387c4d99f435c42370b282fa6/components/driver/mcpwm/mcpwm_private.h#L37) in the IDF. This limits us to a [maximum resolution](https://github.com/espressif/esp-idf/blob/7052a14d61d98dd387c4d99f435c42370b282fa6/components/driver/mcpwm/mcpwm_com.c#L144) of 160MHz/2=80MHz :/

ivmarkov commented:
Yes - will look at it over the weekend. @MabezDev if you can take a look too, that would be appreciated!

MabezDev commented on Cargo.toml:
```diff
@@ -13,6 +13,9 @@ readme = "README.md"
 documentation = "https://esp-rs.github.io/esp-idf-hal/"
 rust-version = "1.61"
 
+[patch.crates-io]
```
Can we remove this patch, there has been a recent release.

MabezDev reviewed (changes requested):
I've tested the example on my s3 board with a logic analyser and the output looks good to me.

I think the number of TODOs is okay and can be resolved later. I think that adding v4.4 support can also be resolved in a follow-up PR, but I'll let @ivmarkov decide whether that should be done here or not.

Just one comment about using a release esp-idf-sys to resolve on my end.

usbalbin commented on Cargo.toml:
```diff
@@ -13,6 +13,9 @@ readme = "README.md"
 documentation = "https://esp-rs.github.io/esp-idf-hal/"
 rust-version = "1.61"
 
+[patch.crates-io]
+esp-idf-sys = { git = "https://github.com/esp-rs/esp-idf-sys", branch = "master" }
+
 [features]
```
```suggestion
[features]
```

usbalbin commented on src/mcpwm/comparator.rs:
```diff
@@ -0,0 +1,127 @@
+    fn default() -> Self {
+        let mut flags: mcpwm_comparator_config_t__bindgen_ty_1 = Default::default();
+        // TODO: What should be set here?
+        flags.set_update_cmp_on_tep(0);
+        flags.set_update_cmp_on_tez(1);
```
Just remembered https://github.com/espressif/esp-idf/issues/9904 . Do you think `tep` be set to `1` to avoid that? I have not tested this PR for that problem yet

usbalbin replied:
```suggestion
        flags.set_update_cmp_on_tep(1);
        flags.set_update_cmp_on_tez(1);
```

ivmarkov commented on src/mcpwm/generator.rs:
```diff
@@ -0,0 +1,307 @@
+    }
+}
+
+pub trait GenInit {
```
`GenInit` has a single implementation. Why are we introducing this trait, then?

ivmarkov commented on src/mcpwm/generator.rs:
```diff
@@ -0,0 +1,307 @@
+
+pub struct NoGenCfg;
+
+pub trait OptionalGenCfg {
```
Rather than introducing `OptionalGenCfg` trait, why not just using `Option<GeneratorConfig>`?

ivmarkov commented on src/mcpwm/generator.rs:
```diff
@@ -0,0 +1,307 @@
+    }
+}
+
+pub trait OnMatchCfg {
```
Ditto: we can probably get rid of this after removing most of the other traits.

ivmarkov commented on src/mcpwm/timer.rs:
```diff
@@ -0,0 +1,302 @@
+}
+
+impl<const N: u8, G: Group> Timer<N, G> {
+    pub fn new(timer: TIMER<N, G>, config: TimerConfig) -> Self {
```
`TimerConfig` should be passed by reference, as we do in all other drivers.

ivmarkov replied:
The `timer` peripheral should be passed as (roughly) `timer: impl Peripheral<P = impl Timer> + 'd` where `Timer` (to be renamed to `TimerDriver`) is parameterized by the `'d` lifetime. The current implementation does not allow for taking a `&mut ref` to the timer peripheral, which means that dropping the driver will always lose the peripheral.

ivmarkov commented on src/mcpwm/timer.rs:
```diff
@@ -0,0 +1,302 @@
+    period_peak: u16,
+}
+
+impl<const N: u8, G: Group> Timer<N, G> {
```
If `Timer` is the driver for the `TIMER` peripheral, it should be named `TimerDriver`. We do this in all other peripherals now.

usbalbin replied:
Same thing for `Comparator`, `Generator` and `Operator`?

ivmarkov commented on src/mcpwm/timer.rs:
```diff
@@ -0,0 +1,302 @@
+    }
+}
+
+pub struct Timer<const N: u8, G: Group> {
```
Having `_group` and `_timer` undescored means you don't actually use them after the `new` constructor. Can we get rid of those? That would mean getting rid of the `N` and `G` generics on the `Timer(Driver)` which simplifies the type.

usbalbin replied:
See [discussion](https://github.com/esp-rs/esp-idf-hal/pull/132#discussion_r1035028499). For(not added in this PR) to be able to support some limited version of this for IDF 4.4 later on, we need to be able to limit what `Timer` can be attached to what `Operator`.

In general a timer can only be connected to operators in the same `Group`, this is a hardware limitation. With IDF 4.4 there is also the constraint that `Timer<X, G>` can only be connect to `Operator<Y, G>` iff `X=Y`. This is done using (misspelled) attatch_operator{A} (soon to be `attach_operator{A}`).

usbalbin replied:
Anyway, I could not come up with a less generics heavy way to express something like that. I am however very open to suggestions :)

ivmarkov replied:
If the `G` generic is remaining, that's still not the end of the world, as long as we can still simplify / remove most of the other generics induced by traits in the code.

ivmarkov commented on src/mcpwm/timer.rs:
```diff
@@ -0,0 +1,302 @@
+    }
+}
+
+pub struct TIMER<const N: u8, G: Group> {
```
I've asked for the removal of as many traits from this impl as possible, but introducing a `Timer` trait here, which is a marker trait for a `TIMER` **peripheral** (don't confuse it with the existing `Timer` struct which should be renamed to `TimerDriver` - see below)), has the big benefit that it will help you to remove a lot of the generics from the driver (`TimerDriver`).

ivmarkov commented on src/mcpwm/timer.rs:
```diff
@@ -0,0 +1,302 @@
+
+impl<const N: u8, G: Group> crate::peripheral::sealed::Sealed for TIMER<N, G> {}
+
+impl<const N: u8, G: Group> crate::peripheral::Peripheral for TIMER<N, G> {
```
Just to note that implementing `Peripheral` only makes sense if you allow - in the `TimerDriver::new` constructor - to pass the peripheral not just by moving, but also by `&mut ref`, which is not possible right now. I'll elaborate further down in the code how that can become possible.

ivmarkov commented on src/mcpwm/timer.rs:
```diff
@@ -0,0 +1,302 @@
+        _timer
+    }*/
+
+    pub fn into_connection(self) -> TimerConnection<N, G, NoOperator, NoOperator, NoOperator> {
```
I have some concerns around this method being consuming, but let's discuss this later, after we have covered some of the other changes I've suggested.

ivmarkov reviewed (changes requested):
@usbalbin @MabezDev I think we might be using way too many traits and generics where simpler types would suffice. As a result, the driver types are very very generics heavy, whereas the HAL crate has been moving in precisely the opposite direction during the last several months and with the latest release of the HAL. I'm also struggling to see the need for most of these traits?

Am I missing something?

usbalbin commented:
Thanks a lot for the review!

usbalbin commented on src/mcpwm/comparator.rs:
```diff
@@ -0,0 +1,127 @@
+pub struct CmpY;
+impl ComparatorChannel for CmpY {}
+
+pub trait OptionalCmp {
```
Again, I am not quite happy with the huge amount of traits either :/

The motivation behind `OptionalCmp` is to only provide a `operator.set_compare_value_x` iff `CMPX` is a `Comparator`. https://github.com/esp-rs/esp-idf-hal/blob/4acbca7a35a6bbae089ccce03c791c4981654037/src/mcpwm/operator.rs#L143

I personally do not quite like a method doing that check at runtime. I will hower change it if you would.

usbalbin replied:
I do not have time the time to check right now but I believe there was also something similar when building the generator config. The generator config should not depend on events from comparators that are not provided.

usbalbin replied:
On second thought, perhaps both comparators should be obligatory since they can not be used for anything else outside their operator anyways...

usbalbin commented:
I have now made the two Comparators of every `Operator` mandatory. That way we get rid two of `Operator`'s type parameters. Same thing could be done for the `Generator`s, however there is, as far as I know, no way to create a generator without assigning a pin. Since not all users need both pins, I would prefer if this was optional. I suppose do that by `Optional<Generator<.., PIN>>`. However is there any good way to not force the user to specify the type of `PIN` when passing `None`?

ivmarkov commented:
> I have now made the two Comparators of every `Operator` mandatory. That way we get rid two of `Operator`'s type parameters.
> 
> Same thing could be done for the `Generator`s, however there is, as far as I know, no way to create a generator without assigning a pin. Since not all users need both pins, I would prefer if this was optional. I suppose do that by `Optional<Generator<.., PIN>>`. However is there any good way to not force the user to specify the type of `PIN` when passing `None`?

`const NO_PIN: Option<AnyIOPin> = None`. And then you can pass NO_PIN everywhere.

ivmarkov commented:
> I have now made the two Comparators of every `Operator` mandatory. That way we get rid two of `Operator`'s type parameters.
> 
> Same thing could be done for the `Generator`s, however there is, as far as I know, no way to create a generator without assigning a pin. Since not all users need both pins, I would prefer if this was optional. I suppose do that by `Optional<Generator<.., PIN>>`. However is there any good way to not force the user to specify the type of `PIN` when passing `None`?

And BTW you can get rid of the `P` generic in your generator struct. I.e. from
```rust
pub struct Generator<G, P: OutputPin>
```

into
```rust
pub struct Generator<G>
```

Why don't you look into how the other drivers do it? As in the SPI and I2C ones? The idea is that only the constructor of the struct is generified by `P` but the struct itself isn't. You just keep the pin number in the struct, not the actual pin peripheral.

ivmarkov commented:
There are also other issues, like the fact that you need to acquaint yourself with the `PeripheralRef` and `pin: impl Peripheral<P = impl OutputPin> + d` syntax. Where I'm getting at, is that your current `Generator` struct in its current shape has also other issues, like the fact that it keeps the pin as `P: OutputPin`, while it should keep it as `PeripheralRef<P> where P: OutputPin`. That is, if it was necessary to keep the peripheral in the `Generationr` struct in the first place, which it isn't.

ivmarkov commented:
There are BTW a bunch of other `Option`-inventing traits throughout the code. I hope you plan to get rid of these as well. If not, let me know, and we can discuss why not (as in what difficulties you see.)

usbalbin commented:
> There are BTW a bunch of other `Option`-inventing traits throughout the code. I hope you plan to get rid of these as well. If not, let me know, and we can discuss why not (as in what difficulties you see.)

Yes, I know, I think I should be able to fix most/all of those now...

> const NO_PIN: Option<AnyIOPin> = None. And then you can pass NO_PIN everywhere.

Oh, that is very nice.

> There are also other issues, like the fact that you need to acquaint yourself with the PeripheralRef and pin: impl Peripheral<P = impl OutputPin> + d syntax. Where I'm getting at, is that your current Generator struct in its current shape has also other issues, like the fact that it keeps the pin as P: OutputPin, while it should keep it as PeripheralRef<P> where P: OutputPin. That is, if it was necessary to keep the peripheral in the Generator struct in the first place, which it isn't.

Thanks, I'll look into that

usbalbin commented on src/mcpwm/operator_config.rs:
```diff
@@ -0,0 +1,121 @@
+        pin_a: PA,
+        pin_b: PB,
+    ) -> OperatorConfig<GeneratorConfig<GenA, PA>, GeneratorConfig<GenB, PB>> {
+        OperatorConfig::empty()
+            .cmp_x(Default::default())
+            .cmp_y(Default::default())
+            .gen_a(GeneratorConfig::active_high(pin_a))
+            .gen_b(GeneratorConfig::active_high(pin_b))
```
Here for
```rust
.gen_a(GeneratorConfig::active_high(pin_a))
```

due to GeneratorConfig being generic GeneratorChannel, [active_high](https://github.com/esp-rs/esp-idf-hal/blob/691337462ddfb4539197e4a7e3d37b200091d472/src/mcpwm/generator.rs#L161) knows that the generator A should only listen to comparator X.

Similarly for `.gen_b(GeneratorConfig::active_high(pin_b))`, it knows that generator A should only listen for comparator Y.

I am not quite sure how to achieve something similar using enums

usbalbin replied:
Do you have any thoughts?

usbalbin replied:
The default configuration is to have it work that way, yes. It is also they way it works in IDF 4.4.

However in IDF 5.x you are free to have each generator listen to events from none, any or all comparators.

---

In a future PR when we add support for IDF 4.4 we need to somehow prevent connecting the wrong comparators to the generator.

usbalbin replied:
> In a future PR when we add support for IDF 4.4 we need to somehow prevent connecting the wrong comparators to the generator.

This is a bit similar to the [discussion](https://github.com/esp-rs/esp-idf-hal/pull/132#discussion_r1035028499) about how operators and timers are related

ivmarkov replied:
How about the following simplification then: promoting the connection of the wrong comparator to the generator we can implement as a runtime error, in the name of simplicity? For ESP IDF 4 it will generate a runtime error, and for ESP IDF 5 it will not?

I'm just afraid that we are generating a very heavy API for something that over time will disappear - i.e. ESP IDF 4.4 specifics.

usbalbin replied:
Makes, sense. And in the end if this is what you prefer than I will try to do that.

However, I do not think that solves my [initial concern](https://github.com/esp-rs/esp-idf-hal/pull/132#discussion_r1089788561). Further, using the same argument, it would mean that we force users to handle/unwrap an error that in time will not happen. 

Again, not quite happy with the current solution either and I really hear you regarding the enormous amount of generics, but it feels a bit sad to push things we know at compile time to runtime.

ivmarkov replied:
Well I'm also not happy with pushing things to runtime, especially given that one of the major premises of Rust is compile time safety.

But this is all a fine-balance act, right? As in, I've come to realize that simplicity of use and simpler types is just as important as compile-type safety.

Let me put it another way: if we do this **one** sacrifice - for ESP-IDF 4 only - would that allow us to get rid of most of the generics?

usbalbin replied:
>But this is all a fine-balance act, right? As in, I've come to realize that simplicity of use and simpler types is just as important as compile-type safety.

Yes, very much so :)

> Let me put it another way: if we do this one sacrifice - for ESP-IDF 4 only - would that allow us to get rid of most of the generics?

Yes, I think atleast GENA/GENB and CMX/CMPY can be removed if we rethink `OperatorConfig` a bit.

usbalbin replied:
I will give it a try :)

ivmarkov replied:
Thank you so much! I know it's been a long ride...

usbalbin replied:
No problem at all, thank *you*!

usbalbin commented:
The main types the user will be passing around as of right now looks like this
```rust
struct TimerDriver<const N: u8, G: Group> {...}
struct Operator<'d, const N: u8, G: Group> {...}
struct TimerConnection<const N: u8, G: Group, O0, O1, O2> {...}
```

I suppose `TimerDriver`(and therefore also `TimerConnection`) will also get a lifetime parameter once i have done the `impl Peripheral<T = Timer>` thing for that too.

What types should get the `Driver` suffix? All of `Operator`, `TimerConnection`, `Generator` and `Comparator`?

ivmarkov commented:
> The main types the user will be passing around as of right now looks like this
> 
> ```rust
> struct TimerDriver<const N: u8, G: Group> {...}
> struct Operator<'d, const N: u8, G: Group> {...}
> struct TimerConnection<const N: u8, G: Group, O0, O1, O2> {...}
> ```
> 
> I suppose `TimerDriver`(and therefore also `TimerConnection`) will also get a lifetime parameter once i have done the `impl Peripheral<T = Timer>` thing for that too.
> 
> What types should get the `Driver` suffix? All of `Operator`, `TimerConnection`, `Generator` and `Comparator`?

For now, let's just have `TimerDriver` suffixed with `*Driver`. Once this lands, we can revisit shortly after that.
Question about the remaining generics:
* Why do we need `N`?
* Why do we still need `O0`, `O1` and `O2`?

ivmarkov commented:
IF that brings additional clarity: we are continuing to (ab)use traits to express patterns where I hoped a simple `Option` would do. As in `OptionalOperator`. Why do we need to express that an operator is optional with a dedicated trait?

usbalbin commented:
> For now, let's just have `TimerDriver` suffixed with `*Driver`. Once this lands, we can revisit shortly after that. [...]

Ok

> * Why do we need `N`?

`TimerConnection` is used to ensure the operators' won't outlive the `Timer`. `TimerConnection::split()` gives us mutable references to both timer and operators. If we allowed multiple instances of the same operator type, the user could use `mem::swap` to move the `Operator` out of the `TimerConnection` which would be bad. Could this be checked at runtime? Maybe, have not given it enough thought..

Other than that, there is also the issue of IDF4 only supporting `Timer<A>` connecting to `Operator<B>` iff A=B.

> * Why do we still need `O0`, `O1` and `O2`?

Thanks to that, `TimerConnection::split()` can return the real types [without the user having to unwrap any `Options`](https://github.com/esp-rs/esp-idf-hal/blob/66d0dde4f1883b08c493b9e6d30409070a709dda/examples/mcpwm-simple.rs#L83). I was also considering adding `TimerConnection::timer()` and `TimerConnection::operator{N}` methods at some point(nice to have when only needing access to one object at a time). `TimerConnection::operator{N}` would only be implemented if `O{N}` is a real `Operator`. In my mind this would be really nice for discoverability when using something like vs code + rust analyzer or similar.

Is it worth the three type parameters? Not sure... :)

noahbliss commented:
Hey @usbalbin I'm looking to implement/leverage this as a pwm capture method, any pointers on where you left off?

usbalbin commented:
> Hey @usbalbin I'm looking to implement/leverage this as a pwm capture method, any pointers on where you left off?

I have not really looked at the capture thing. However most of the types so far are quite close to how the diagram are drawn [here](https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/peripherals/mcpwm.html), so I would probably start there and look through the diagrams and functions/structures to figure out how things relate.

From a quick glance, the capture thing seems to be quite independent having its own internal timer etc, yet it seems it can be synced with other timers from the same mcpwm unit.

As to the state of the code here, I do not quite remember. I have not worked on this for quite a while. I believe it should work but there is likely a lot of room for improvement :)

usbalbin commented:
I am not working at this right now, and I do not see myself doing so anytime soon. So feel free to use this/finish it as you wish :)

mindoodoo commented:
Hey, I'd very much try to use this as I need the new MCPWM api (5.x) for my application.

I tried making `esp-idf-hal` a git dep to your fork, however am running into build issues where it seems like the bindings have changed since you last worked on this. Would you maybe know know which version of ESP-IDF your bindings were getting compiled from and how I could change the target esp idf version used to create the bindings ?

usbalbin commented:
It was some quite some time since I did anything with this...

However, digging through my files:
.cargo/config.toml:
```toml
...
[env]
ESP_IDF_VERSION = { value = "branch:release/v5.0" }
```

Cargo.toml:
```toml
...
#esp-idf-svc = { version = "0.45.0" } # <--- Not sure if this is what I based my local version on?
esp-idf-svc = { path = "..." }
esp-idf-sys = { version = "0.32.1", default-features = false, features = ["binstart", "panic_handler", "alloc_handler"] }
...
```

I have changed mcu since, so no guarantees, but I believe this worked when it was written.
